### ISSUE TITLE
If 1080Ti GPU Fully Used

### ISSUE THREAD

**Issue by wxthss82** [ageitgey/face_recognition#354]
* face_recognition version:latest
* Python version:3.5.2
* Operating System:Ubuntu 16.04.3
* Webcam 8M WebCam, 3264*2448

### Description

I build dlib with CUDA support, I try to use the facerec_from_webcam.py, but I find a low gpu usage,
I pass model = "cnn", still low usage.

### What I Did
![screenshot from 2018-02-09 15-40-42](https://user-images.githubusercontent.com/3448682/36016921-608ac2be-0db0-11e8-94c1-68848f0c3c74.png)

**Comment by ageitgey** [comment]
What's the problem exactly? Is the performance less than you need so you want it to use more GPU resources or something?

**Comment by facepainter** [comment]
facerec_from_webcam isn't really going to stress anything as it is still only processing one frame at a time  - try running the find_faces_in_batches.py with a batch size of 128 against an input video.

**Comment by X-YI** [comment]
@ageitgey Yes, I think if the processing time is fast enough, It can do realtime face recognition, but now I see it's about one second delay

**Comment by X-YI** [comment]
@facepainter Thank you for your advice, I think that if the processing time is fast enough, it can be realtime, but now I see about one second delay. That's to say, if it takes 0.099 second to recognize face in picture sequences, and take 0.001 second to grab image file from webcam, it can be about 10fps so the image sequences show in the windows can be continuous, but now I see it as discrete.

**Comment by snuzxj** [comment]
Similar issue here. After some digging, found that dlib's face detection does NOT support GPU nor multi-thread.

**Comment by ageitgey** [comment]
That's not correct.

dlib's non-CNN based face detector doesn't support the GPU but you can easily run it in multiple processes in python (for example, see concurrent.futures) to take advantage of multiple cpu cores.

dlib's CNN-based face detector can use a GPU. Check out some of the demos, like the batch face detection demo.

**Comment by snuzxj** [comment]
Can you show a pointer to dlib's source code which indicates it uses GPU for face detection?

**Comment by snuzxj** [comment]
I am running dlib on Jetson TX2, which has a decent GPU but very weak CPU (which are ARM). It is really slow no mater using HOG or CNN based face detector. And, the face comparison speed is also slower than my MBP with has no GPU... BTW, I set GPU support when compiling and installing dlib.

**Comment by ageitgey** [comment]
@snuzxj The python wrapper for the CNN face detector is [here](https://github.com/davisking/dlib/blob/0f51dfb96e5a2697e932eb7062b6d291977ff65a/tools/python/src/cnn_face_detector.cpp). I wrote the first version of the wrapper in dlib. It's hard to see that it uses the GPU because it's loading a serialized copy of the neural network so you can't easily see how it's defined.

So check out the [face detector C++ example](https://github.com/davisking/dlib/blob/master/examples/dnn_mmod_face_detection_ex.cpp) instead where it's easier to see.  The face detector model is defined here in code using the DNN functions provided by dlib.

The [DNN code](https://github.com/davisking/dlib/tree/master/dlib/dnn) is full of CUDA references. See for example all the `#ifdef DLIB_USE_CUDA` sections in https://github.com/davisking/dlib/blob/master/dlib/dnn/tensor_tools.cpp which the face detector uses several functions from.

**Comment by snuzxj** [comment]
@ageitgey Thank you for the detailed pointers. I will look into this, and see what are the problems. I just came across this thread, which also says dlib doesn't do GPU face detection.
https://github.com/cmusatyalab/openface/issues/157

**Comment by ageitgey** [comment]
That was true in the past when that thread was posted. The GPU model was only available via C++. That's why I contributed a python wrapper to expose that functionality in dlib a few months ago :)

**Comment by deepa042008** [comment]
I have  i5 8350u cpu 64 bit;8GB ram Windows 10 ; python 3.8 with conda 4.7
I HOG model is sort of ok although slow with 5 picture and 5 min video 

I need to capture the video frames of incoming or outgoing employees of a company & capture their in time & out time

So this camera will be continuously running and capturing

1. Should i go for Raspberry Pi . I understand
Should i go for CNN or Tensorflow/Keras 
I would like to use CNN since it is more accurate

I have  i5 8350u cpu 64 bit;8GB ram Windows 10 ; python 3.8 with conda 4.7

I used HOG model is sort of ok although slow with 5 min video . For each face i have only 1 photo to recognize and my known faces are only 2 
I need to capture the video frames of incoming or outgoing employees of a company & capture their in time & out time
So the camera will be continuously running and capturing
Wit CNN model it is extremely slow. 

1. Should i go for Raspberry Pi . I understand from Adrian Rosebrock site that it is not suitable for face_recognition 

2. Should i go for CNN or Tensorflow/Keras or any other model should be tried pls advise

3. I would like to find how to know if GPU is being used. i have followed the a site and done the cmake build of dlib
 I used https://github.com/ageitgey/face_recognition/commit/b17901840000e 
RUN cmake -H/dlib -B/dlib/build -DDLIB_USE_CUDA=1 -DUSE_AVX_INSTRUCTIONS=1
to build dlib
before this i did conda install of dlib
4.  when i execute python setup.py install --yes USE_AVX_INSTRUCTIONS --yes DLIB_USE_CUDA -G "Visual Studio 15 2017 Win64"
it says --yes is not recognized 
5. Also i have installed Microsoft Visual Studio 2017 and not 2015. I understand from https://github.com/ageitgey/face_recognition/issues/303
that CUDA do not support vs2017 & we need t install VS2015. Is there any workaround

6.  how to use face_recognition_svm.py for training.
